Added new exercise 'diffie-hellman'

lathspell opened the pull request.

dector commented on exercises/diffie-hellman/build.gradle:
```diff
@@ -0,0 +1,28 @@
+buildscript {
+  ext.kotlin_version = '1.2.61'
```
Please, update to `1.3.50`.

lathspell replied:
done

lathspell commented:
> I think we also don't need .keep file and need stub source file instead.
also done

dector commented on exercises/diffie-hellman/src/test/kotlin/DiffieHellmanTest.kt:
```diff
@@ -0,0 +1,79 @@
+        }
+    }
+
+    // Note: due to the nature of randomness, there is always a chance that this test fails
```
Let's make follow-up issue to improve this in future.

dector commented on exercises/diffie-hellman/src/test/kotlin/DiffieHellmanTest.kt:
```diff
@@ -0,0 +1,79 @@
+    // Be sure to check the actual generated values
+
+    @Test
+    fun testPrivateKeyIsRandom() {
```
Let's use back-ticks for test names.

dector commented on exercises/diffie-hellman/src/main/kotlin/DiffieHellman.kt:
```diff
@@ -0,0 +1,3 @@
+class DiffieHellman {
+    
```
Let's add required functions with `TODO()` body. Stubs are required now.

dector commented on exercises/diffie-hellman/src/main/kotlin/DiffieHellman.kt:
```diff
@@ -0,0 +1,3 @@
+class DiffieHellman {
```
Do we actually need class here? Seems like `object` will be enough for now.

eparovyshnaya replied:
We already have almost a half of exercises implemented by  `object`. Let's start [reducing this amount](https://github.com/exercism/kotlin/pull/339#discussion_r345259383) (, please!). I'd rather recomment to treat a `class`-based implementation as a default one.

dector replied:
I don't see a need for state-less classes here, generally. It's not an idiomatic Kotlin. :) Top-level function is better but some use `object`s as namespaces (since Kotlin don't have namespaces).

eparovyshnaya commented on exercises/diffie-hellman/README.md:
```diff
@@ -0,0 +1,69 @@
+
+Diffie-Hellman key exchange.
+
+Alice and Bob use Diffie-Hellman key exchange to share secrets.  They
```
Would it be more informative to place a trusted long-living reference to a detaild description here? 

Like [Diffie-Hellman key exchange](https://en.wikipedia.org/wiki/Diffie%E2%80%93Hellman_key_exchange)

eparovyshnaya commented on exercises/diffie-hellman/src/test/kotlin/DiffieHellmanTest.kt:
```diff
@@ -0,0 +1,79 @@
+    }
+
+    // Note: due to the nature of randomness, there is always a chance that this test fails
+    // Be sure to check the actual generated values
```
Can we form this note as a regular kdoc-comment?

dector replied:
I guess it's better to make a follow-up issue to fix this later. Tests should introduce unexpected behavior never. :) But this more a question to specification.

lathspell commented:
- changed class -> object
- added TODO() stub methods
- reformatted comments to KDoc
- the link to Wikipedia was already present in the README, just a bit further down

eparovyshnaya commented on exercises/diffie-hellman/src/test/kotlin/DiffieHellmanTest.kt:
```diff
@@ -0,0 +1,79 @@
+
+    @Before
+    fun setUp() {
+        diffieHellman = DiffieHellman()
```
After you reimplemented `DiffieHellman` to `object`, the test should accomodate new signature, as `DiffieHellman()` can no longer be treated as ctor.

dector commented on exercises/diffie-hellman/.meta/src/reference/kotlin/DiffieHellman.kt:
```diff
@@ -0,0 +1,22 @@
+import java.math.BigInteger
+import java.util.Random
+
+class DiffieHellman {
```
Oops. `object` needed here.

dector commented:
Please, fix `class` -> `object` and we can merge this, I guess.

eparovyshnaya commented:
You still need to fix Test to achieve compilability. You cannot instanciate `object`.

dector commented on exercises/diffie-hellman/src/test/kotlin/DiffieHellmanTest.kt:
```diff
@@ -0,0 +1,78 @@
+
+class DiffieHellmanTest {
+
+    private lateinit var diffieHellman: DiffieHellman
```
There is no need for constructing `object` as it's singleton already. :)

eparovyshnaya commented:
@lathspell 
Here is a patch for test suite and stub implementation with the following content: 
 - `object` utilization is fixed across the test suite
 - tests are renamed with backtics
 - *reason* is added to `TODO` calls in stub implementation

[tests_and_stub.patch](https://github.com/exercism/kotlin/files/3863638/tests_and_stub.patch.txt)
It is not signed-off, so you can modify anything on top of it. 

Consider appling it to your branch, it's going to resolve all open conversations here. 

Prior to push, please [squash all commits in the branch](https://github.com/exercism/docs/blob/master/contributing/git-basics.md#squashing).

eparovyshnaya commented:
Dear @lathspell, it's everithing ok with your exercise now.
But since the time of this PR opening, all our exercises builds migrated to kotlin dsl, updated gradle and kotlin versions. 

Have a look at #358. Currently we use 
 - gradle of 6.0.1 version
 - kotlin 1.3.60
 - gradle.build is turned to gradle.build.kts

Please, update your exercise according to these chages, unless our Travis will keep failing. 

I truely thank you for your work. Let's do this last step to get your exercise merged.

lathspell commented:
> nit: Usually we mark all tests as ignored (except first one). But it's legacy approach that we are discussing now.

I know and I hated it that I had to delete those @ Ignore on every excercise I made so I hope you decide against making them mandatory :)

Thanks for merging!
